### PR TITLE
Wasm update overview and interop

### PR DESCRIPTION
**Pull request by AlejandraPedroza**
Updated wasm-overview.md and wasm-js-interop.md to make them consistent with Wasm in Alpha.

### CODE REVIEW

File: docs/topics/wasm/wasm-overview.md [hunk 1]
```diff
@@ -1,61 +1,75 @@
...
-> We would appreciate your feedback on it in [YouTrack](https://kotl.in/issue).
+> [Join the Kotlin/Wasm community](https://slack-chats.kotlinlang.org/c/webassembly).
 >
 {type="note"}
 
-[WebAssembly (Wasm)](https://webassembly.org) is a binary instruction format for a stack-based virtual machine.
-This format is platform-independent because it runs on its own virtual machine.
-Wasm is designed to be fast and secure, and it can compile code from various programming languages, including Kotlin.
+Kotlin has the power to build applications, and reuse mobile and desktop user interfaces (UIs) in your web projects with
+Compose Multiplatform for Web and Kotlin/Wasm.
```
Comment (danil-pavlov) [inline docs/topics/wasm/wasm-overview.md:11]:
```suggestion
Compose Multiplatform and Kotlin/Wasm.
```

File: docs/topics/wasm/wasm-overview.md [hunk 2]
```diff
@@ -1,61 +1,75 @@
...
 
-Kotlin/Wasm is a new compilation target for Kotlin. You can use it in your Kotlin Multiplatform projects.
-With Kotlin/Wasm, you can create applications that run on different environments and devices supporting WebAssembly and meeting Kotlin's requirements.
+[Compose Multiplatform]((https://www.jetbrains.com/lp/compose-multiplatform/)) is a declarative framework based on Kotlin
+and [Jetpack Compose](https://developer.android.com/jetpack/compose), which allows you to implement the UI
+once and share it across all the platforms you target. Specifically for web platforms, Compose Multiplatform for Web uses
```
Comment (danil-pavlov) [inline docs/topics/wasm/wasm-overview.md:15]:
```suggestion
once and share it across all the platforms you target. Specifically for web platforms, Compose Multiplatform uses 
```

File: docs/topics/wasm/wasm-overview.md [hunk 3]
```diff
@@ -1,61 +1,75 @@
...
+## Leave feedback
 
-You can use the Kotlin standard library (`stdlib`) and test library ([`kotlin.test`](https://kotlinlang.org/api/latest/kotlin.test/))
-in Kotlin/Wasm out of the box. The version of these libraries is the same as the version of the `kotlin-multiplatform` plugin.
+### Kotlin/Wasm
+* ![Slack](slack.svg){width=25}{type="joined"} Slack: provide your feedback directly to developers in our [#webassembly](https://kotlinlang.slack.com/archives/CDFP59223) channel. [Get a Slack invite here](https://surveys.jetbrains.com/s3/kotlin-slack-sign-up).
+* ![Youtrack](youtrack-black-logo.png){width=25}{type="joined"} Youtrack: report any issues in [YouTrack](https://youtrack.jetbrains.com/issue/KT-56492).
 
-Kotlin/Wasm has an experimental support for other Kotlin libraries. [Read more how to enable them in your project](wasm-libraries.md).
+### Compose Multiplatform for Web
```
Comment (danil-pavlov) [inline docs/topics/wasm/wasm-overview.md:67]:
```suggestion
### Compose Multiplatform
```

File: docs/topics/wasm/wasm-overview.md [hunk 4]
```diff
@@ -1,61 +1,75 @@
...
+and [Jetpack Compose](https://developer.android.com/jetpack/compose), which allows you to implement the UI
+once and share it across all the platforms you target. Specifically for web platforms, Compose Multiplatform for Web uses 
+Kotlin/Wasm as its compilation target.
 
-> Learn more about Kotlin/Wasm with this [YouTube playlist](https://kotl.in/wasm-pl).
+Explore our online demo of an application built with Compose Multiplatform for Web and Kotlin/Wasm:
+
+<a href="https://zal.im/wasm/jetsnack/"><img src="wasm-demo.png" width="700" alt="Kotlin/Wasm demo"/></a>
+
+>To run applications built with Kotlin/Wasm in a browser, you need a browser version that supports the new garbage collection
```
Comment (koshachy) [inline docs/topics/wasm/wasm-overview.md:22]:
```suggestion
> To run applications built with Kotlin/Wasm in a browser, you need a browser version that supports the new garbage collection
```

File: docs/topics/wasm/wasm-overview.md [hunk 5]
```diff
@@ -1,61 +1,75 @@
...
+> and exception handling proposals. To check the support status, see the [WebAssembly
+> roadmap](https://webassembly.org/roadmap/).
 >
-{type="note"}
+{type="tip"}
 
-## Browser support
+[WebAssembly (Wasm)](https://webassembly.org/) is a binary instruction format for a stack-based virtual machine. This
+format is platform-independent because it runs on its own virtual machine. Wasm provides languages like Kotlin, and many 
+more, with a compilation target so that they can run on the web.
```
Comment (koshachy) [inline docs/topics/wasm/wasm-overview.md:30]: I found this one (the last sentence) a bit hard to read. 
But I'm not sure, maybe it is only me. WDYT about it?

File: docs/topics/wasm/wasm-overview.md [hunk 6]
```diff
@@ -1,61 +1,75 @@
...
-> We would appreciate your feedback on it in the public Slack channel [#compose-web](https://slack-chats.kotlinlang.org/c/compose-web).
-> If you face any issues, please report them on [GitHub](https://github.com/JetBrains/compose-multiplatform/issues).
->
-{type="note"}
+We regularly run benchmarks on Kotlin/Wasm, and these results come from our testing in a recent version of Google Chrome.
+
+## Browser API support
+
+The Kotlin/Wasm standard library provides declarations for browser APIs (including the DOM API).
+With these declarations, you can directly use the Kotlin API to access and utilize various browser functionalities, like
```
Comment (koshachy) [inline docs/topics/wasm/wasm-overview.md:51]: Well, I'd like to see this one a bit simplified. Maybe we should divide it into two parts?

File: docs/topics/wasm/wasm-overview.md [hunk 7]
```diff
@@ -1,61 +1,75 @@
...
-in Kotlin/Wasm out of the box. The version of these libraries is the same as the version of the `kotlin-multiplatform` plugin.
+### Kotlin/Wasm
+* ![Slack](slack.svg){width=25}{type="joined"} Slack: provide your feedback directly to developers in our [#webassembly](https://kotlinlang.slack.com/archives/CDFP59223) channel. [Get a Slack invite here](https://surveys.jetbrains.com/s3/kotlin-slack-sign-up).
+* ![Youtrack](youtrack-black-logo.png){width=25}{type="joined"} Youtrack: report any issues in [YouTrack](https://youtrack.jetbrains.com/issue/KT-56492).
 
-Kotlin/Wasm has an experimental support for other Kotlin libraries. [Read more how to enable them in your project](wasm-libraries.md).
+### Compose Multiplatform for Web
+* ![Slack](slack.svg){width=25}{type="joined"} Slack: provide your feedback in the [#compose-web](https://slack-chats.kotlinlang.org/c/compose-web) public channel.
+* ![Github](git-hub.svg){width=25}{type="joined"} Github: report any issues
+  [here](https://github.com/JetBrains/compose-multiplatform/issues).
```
Comment (koshachy) [inline docs/topics/wasm/wasm-overview.md:70]: We don't use "here" as a part of the link. Maybe it's better to use "report any issues" as a link?

File: docs/topics/wasm/wasm-overview.md [hunk 8]
```diff
@@ -1,61 +1,75 @@
...
 
-Kotlin/Wasm allows you to both use JavaScript code and Browser API from Kotlin, and Kotlin code from JavaScript.
+## Performance
 
-[Learn more about Kotlin Wasm interoperability with JavaScript](wasm-js-interop.md).
+Although Kotlin/Wasm is still in Alpha, Compose Multiplatform running on Kotlin/Wasm already shows encouraging performance 
+traits. You can see that the execution speed outperforms JS and is getting close to JVM:
 
-## Compose Multiplatform for Web
+![Kotlin/Wasm performance](wasm-performance-compose.png){width=700}
```
Comment (koshachy) [inline docs/topics/wasm/wasm-overview.md:44]: It's not clear what is on the Y axis... Can we ask to update the chart and add the data/legend?

File: docs/topics/wasm/wasm-js-interop.md [hunk 9]
```diff
@@ -1,164 +1,227 @@
 [//]: # (title: Interoperability with JavaScript)
 
-Kotlin/Wasm allows you to both use JavaScript code from Kotlin and Kotlin code from JavaScript.
+Kotlin/Wasm allows you to use both JavaScript code in Kotlin and Kotlin code in JavaScript.
 
-The [Kotlin/JS compiler](js-overview.md) already provides the ability to transpile your Kotlin code to JavaScript. The
-Kotlin/Wasm interoperability with JavaScript is designed in a similar way, taking into account that JavaScript is a
-dynamically typed language compared to Kotlin. Follow our guide to configure interoperability in your projects.
+As with [Kotlin/JS](js-overview.md), the Kotlin/Wasm compiler also has interoperability with JavaScript. If you are 
+familiar with Kotlin/JS interoperability, you will find that Kotlin/Wasm interoperability is similar. However,
```
Comment (koshachy) [inline docs/topics/wasm/wasm-js-interop.md:6]: Let's rewrite it to not use the future tense.

File: docs/topics/wasm/wasm-js-interop.md [hunk 10]
```diff
@@ -1,164 +1,227 @@
...
+Consider this JavaScript function: 
 
 ```javascript
-// JavaScript
+function greet(name) {
+    console.log("Hello, " + name + "!");
+}
+```
+
+You can declare it in Kotlin with the `external()` function:
```
Comment (koshachy) [inline docs/topics/wasm/wasm-js-interop.md:34]: Not sure that we nee parenthesis in this case. It is not the name of the function but the name of the keyword, am I right?

Comment (koshachy) [reply docs/topics/wasm/wasm-js-interop.md:34]: If so, the statement "with the `external()` function" is not correct, it's better to say "declare it in Kotlin as an `external` function:"

File: docs/topics/wasm/wasm-js-interop.md [hunk 11]
```diff
@@ -1,164 +1,227 @@
...
-> Some "external" Kotlin/JS features are not supported in Kotlin/Wasm:
-> * Implementing or extending external types
-> * External [enum classes](enum-classes.md)
->
-{type="note"}
 
-### @JsFun annotation
+#### External objects
+
+Consider this JavaScript's variables holding an object:
```
Comment (koshachy) [inline docs/topics/wasm/wasm-js-interop.md:120]: Not sure that we need to use this apostrophe. We usually say "Kotlin variable", "Python variable" but not Kotlin's, Python's. 

It is also can be re-written as "variable in Javascript that holds and object".

**Review by koshachy — CHANGES_REQUESTED** [review]
Hey @AlejandraPedroza,
Huge work!
I've read and left some comments.

Also, pay special attention to:
* Code formatting
* Using "we" instead of "you"

File: docs/topics/wasm/wasm-js-interop.md (hunk 9 above)
Comment (AlejandraPedroza) [reply docs/topics/wasm/wasm-js-interop.md:6]: Changed it for "you can notice".

File: docs/topics/wasm/wasm-js-interop.md (hunk 10 above)
Comment (AlejandraPedroza) [reply docs/topics/wasm/wasm-js-interop.md:34]: That's true. The function is greet().

File: docs/topics/wasm/wasm-js-interop.md [hunk 12]
```diff
@@ -1,164 +1,227 @@
...
 ```
 
-See the full code in the example
-project [Kotlin/Wasm browser](https://github.com/Kotlin/kotlin-wasm-examples/tree/main/browser-example).
+External interfaces don't have runtime type information and are a compile-time-only concept.
+Therefore, external interfaces have some restrictions compared to regular interfaces:
+* You can't use them on the right-hand side of `is` checks.
+* You can't use them in class literal expressions (such as `User::class`).
+* You can't pass them as reified type arguments.
+* `as` casts to external interfaces always succeed.
```
[the remark this answers is not in the document]
Comment (AlejandraPedroza) [reply docs/topics/wasm/wasm-js-interop.md:115]: Changed it to "Casting with `as` to external interfaces". Is it more understandable?

File: docs/topics/wasm/wasm-js-interop.md [hunk 13]
```diff
@@ -167,137 +230,145 @@ functions:
...
-fun jsTypeOf(obj: Any): String {
-    return js("typeof obj")
-}
+## Exception handling
 
-// Kotlin/Wasm
-@JsFun("(obj) => typeof obj")
-external fun jsTypeOf(obj: SomeExternalInterfaceType): String
-```
+The Kotlin/Wasm `try-catch` expression can't catch JavaScript exceptions.
```
[the remark this answers is not in the document]
Comment (AlejandraPedroza) [reply docs/topics/wasm/wasm-js-interop.md:332]: I just saw it in Kotlin examples, not necessarily Kotlin/Wasm, but I'm doublechecking with SME.

File: docs/topics/wasm/wasm-overview.md (hunk 5 above)
Comment (AlejandraPedroza) [reply docs/topics/wasm/wasm-overview.md:30]: Changed it to "Wasm provides Kotlin and other languages with a compilation target to run on the web". Clearer?

File: docs/topics/wasm/wasm-overview.md (hunk 8 above)
Comment (AlejandraPedroza) [reply docs/topics/wasm/wasm-overview.md:44]: Checking with SME to find who created it and see if we can update it.

File: docs/topics/wasm/wasm-js-interop.md [hunk 14]
```diff
@@ -1,164 +1,227 @@
...
+
+All declarations inside the `external` class are implicitly considered external.
 
-// Use external functions to call JS functions defined in global scope
-external fun consoleLogExample(): Unit
+#### External interfaces
 
-// In addition to functions, you can have external top-level properties
-external var externalInt: Int
+You can describe the shape of a JavaScript object in Kotlin. Consider this JavaScript function and what it returns:
```
[the remark this answers is not in the document]
Comment (AlejandraPedroza) [reply docs/topics/wasm/wasm-js-interop.md:91]: I had discussed this with the SME, and we need to highlight what it returns, that's why in this case it comes with a 2nd sentence keeping consistency with others introducing code blocks. 

Regarding shape, I double-checked with the SME. He says "shape" is fine given that are equivalent to TypeScript interfaces, described with shapes: https://www.typescriptlang.org/docs/handbook/interfaces.html

WDYT?

File: docs/topics/wasm/wasm-js-interop.md [hunk 15]
```diff
@@ -1,164 +1,227 @@
...
+```
 
-To include a small piece of JS code in your Kotlin/Wasm module, use the `@JsFun` annotation with external top-level
-functions. The annotation argument should be a string with JS code that evaluates a function with a matching signature:
+If you want to run a block of JavaScript statements, surround your code inside the string with curly brackets `{}`:
 
 ```kotlin
-@JsFun("function count(x) { return x + 10; }")
-external fun count(x: Int): Int
+fun setLocalSettings(value: String): Unit = js("""{
```
[the remark this answers is not in the document]
Comment (AlejandraPedroza) [reply docs/topics/wasm/wasm-js-interop.md:159]: I reformatted this and the rest of the code blocks.

File: docs/topics/wasm/wasm-js-interop.md [hunk 16]
```diff
@@ -167,137 +230,145 @@ functions:
...
+| `Byte`, `Short`, `Int`, `Char`              | `Number`                          |
+| `Float`, `Double`,                          | `Number`                          |
+| `Long`,                                     | `BigInt`                          |
+| `Boolean`,                                  | `Boolean`                         |
+| `String`,                                   | `String`                          |
+| `Unit` in return position                   | `undefined`                       |
+| Function type, for example `(String) → Int` | Function                          |
+| `JsAny` and subtypes                        | Any JavaScript value              |
+| `JsReference`                               | Opaque reference to Kotlin object |
+| Other types                                 | Not supported                     |
```
[the remark this answers is not in the document]
Comment (AlejandraPedroza) [reply docs/topics/wasm/wasm-js-interop.md:261]: I double-checked with SME and he says we don’t support JsExport on types yet, not even with the annotation.

File: docs/topics/wasm/wasm-js-interop.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (AlejandraPedroza) [reply docs/topics/wasm/wasm-js-interop.md:332]: Confirmed: only Kotlin expression.

File: docs/topics/wasm/wasm-overview.md (hunk 8 above)
Comment (AlejandraPedroza) [reply docs/topics/wasm/wasm-overview.md:44]: Changed (:

File: docs/topics/wasm/wasm-js-interop.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (koshachy) [reply docs/topics/wasm/wasm-js-interop.md:95]: OK

**Review by koshachy — APPROVED** [review]

**Pull request merged**